### ISSUE TITLE
ConvTranspose backwards with dilation and output padding segfaults

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#5112]
UPDATE: ASAN reports a problem with the following test case:

```
import torch
import torch.nn as nn
from torch.autograd import Variable

model = nn.ConvTranspose2d(4, 1, 3, output_padding=1, dilation=2, bias=False)
model.weight.data.fill_(1)
i = Variable(torch.ones(1, 4, 1, 1))
o = model(i)
o.sum().backward()
```

(So, it looks like we're not checking `output_padding < stride`.)

----

The following program reliably segfaults for me:

```
import torch
from torch import nn
from torch.autograd import Variable

batch_size = 1
x = Variable(torch.randn(batch_size, 3, 32, 32))

def Lop(ys, xs, ws):
    return torch.autograd.grad(ys, xs, grad_outputs=ws, create_graph=True, retain_graph=True, allow_unused=True)

def Rop(ys, xs, vs):
    if isinstance(ys, tuple):
        ws = [Variable(torch.zeros(y.size()), requires_grad=True) for y in ys]
    else:
        ws = Variable(torch.zeros(ys.size()), requires_grad=True)

    gs = torch.autograd.grad(ys, xs, grad_outputs=ws, create_graph=True, retain_graph=True, allow_unused=True)
    re = torch.autograd.grad(gs, ws, grad_outputs=vs, create_graph=True, retain_graph=True, allow_unused=True)
    return re

class Test(nn.Module):
    def __init__(self):
        super(Test, self).__init__()
        self.conv1 = nn.Conv2d(3, 16, kernel_size=3)
        self.conv2 = nn.Conv2d(16, 160, kernel_size=3)
        self.conv3 = nn.Conv2d(160, 160, kernel_size=1)
        self.conv4 = nn.Conv2d(16, 160, kernel_size=3)
        self.conv5 = nn.Conv2d(160, 160, kernel_size=1)
        self.conv6 = nn.Conv2d(160, 160, kernel_size=1)
        self.conv7 = nn.Conv2d(160, 320, kernel_size=1, stride=2)
        self.conv8 = nn.Conv2d(320, 320, kernel_size=1)
        self.conv9 = nn.Conv2d(160, 320, kernel_size=1, stride=2)
        self.linear = nn.Linear(62720, 10, bias=False)
    def forward(self, t0):
        t53 = self.conv1(t0)
        t63 = self.conv2(t53)
        t73 = self.conv3(t63)
        t77 = self.conv4(t53)
        t78 = t77 + t73

        t88 = self.conv5(t78)
        t98 = self.conv6(t88)
        t99 = t78 + t98

        t109 = self.conv7(t99)
        t119 = self.conv8(t109)
        t123 = self.conv9(t99)
        t124 = t123 + t119

        x = t124
        x = x.view(x.size(0), -1)
        x = self.linear(x)
        return x

m = Test()
y = m(x)
l = y.sum()

xs = list(m.parameters())
ds = torch.autograd.grad(l, xs, create_graph=True, retain_graph=True)

vs = [Variable(torch.randn(g.size())) for g in ds]
Rop(ds, xs, vs)
"""
# Correctness check
for r1, r2 in zip(Lop(ds, xs, vs), Rop(ds, xs, vs)):
    print((r2-r1).abs().sum())
"""
```

Looks like an engine problem.

CC @apaszke @colesbury 

EDIT: updated with smaller repro
EDIT 2: restored original repro, it seems to trigger problems more reliably

**Comment by ezyang** [comment]
This module also reproduces, but less reliably:

```
class Test(nn.Module):
    def __init__(self):
        super(Test, self).__init__()
        self.conv4 = nn.Conv2d(16, 160, kernel_size=3)
        self.conv7 = nn.Conv2d(160, 320, kernel_size=1, stride=2)
        self.conv8 = nn.Conv2d(320, 320, kernel_size=1)
        self.conv9 = nn.Conv2d(160, 320, kernel_size=1, stride=2)
        self.linear = nn.Linear(72000, 10, bias=False)
    def forward(self, t53):
        t99 = self.conv4(t53)

        t109 = self.conv7(t99)
        t119 = self.conv8(t109)
        t123 = self.conv9(t99)
        t124 = t123 + t119

        x = t124
        x = x.view(x.size(0), -1)
        x = self.linear(x)
        return x
```

**Comment by ezyang** [comment]
Bisected to:

```
commit a88a8ec8278e19f52cfd6e75a685ce0e9200b96b
Author: Edward Z. Yang <ezyang@mit.edu>
Date:   Wed Dec 20 14:19:27 2017 -0500

    Convolution derivatives in ATen (#4116)
    
    * Convolution derivatives in ATen
    
    This PR introduces ATen implementation of convolution, which dispatches to
    THNN/CuDNN/nnpack based on input parameters. The general strategy is to compose
    this function out of the various forward-backward pairs of specific
    implementations, rather than write a monolithic function with backwards (which
    is what we did before because the boilerplate of doing it otherwise would have
    been very high.) The new API provides the following functions:
    
      - _convolution, which is a fully generic, native convolution implementation
        that dispatches to various other convolution implementations depending on
        input characteristics. This is prefixed with an underscore because it
        explicitly takes benchmark, deterministic and cudnn_enabled which are
        implementation details for CuDNN. The intent is to eventually provide a
        convolution that reads these parameters out of the context using #4104.
      - _convolution_nogroup is a convolution implementation for non-CuDNN
        algorithms which don't support group convolution natively.
      - _convolution_double_backward is the generic double-backwards implementation
        for convolution.
```

**Comment by ezyang** [comment]
CPU started segfaulting in:

```
Author: Tongzhou Wang <SsnL@users.noreply.github.com>
Date:   Fri Feb 2 12:39:33 2018 -0500

    Fix output_nr not incremented correctly (#4812)
    
    * fix output_nr not incremented correctly
    
    * update test_conv_double_backward to cover this case; call accGradParameters if any param (not just weight) requires grad in parse_nn.py
    
    * update Spatial/VolumetricFull(Dilated)Convolution to support accGradParameters with only bias requiring grad
    
    * Spatial/VolumetricConvolutionMM
    
    * Spatial/VolumetricDilatedConvolution
    
    * address @fmassa 's comments
```

**Comment by ezyang** [comment]
On CUDA, there is not a segfault, but the correctness check fails. This repros on f23feca681c5066c70f0fe1516fc2e269d615e93

```
(/home/ezyang/Dev/pytorch-env) [ezyang@devgpu005.ash6 ~/Dev/pytorch] python corrupt.py                                  
Variable containing:
1.00000e-06 *
  3.0994
[torch.cuda.FloatTensor of size (1,) (GPU 0)]
        
Variable containing:
 6.5500
[torch.cuda.FloatTensor of size (1,) (GPU 0)]

Variable containing:
 6.5500
[torch.cuda.FloatTensor of size (1,) (GPU 0)]

Variable containing:
 0
[torch.cuda.FloatTensor of size (1,) (GPU 0)]

```

**Comment by ezyang** [comment]
We merged the fix, tests are still coming.
